File: src/VtkIO.cpp
#include "VtkIO.h"
#include <cstdlib>
#include <cassert>

//----------------------------------------------------------------------------

cigma::VtkIO::VtkIO()
{
    fp = NULL;
}

cigma::VtkIO::~VtkIO()
{
}

//----------------------------------------------------------------------------

void cigma::VtkIO::write_header()
{
    fprintf(fp, "# vtk DataFile Version 3.0\n");
    fprintf(fp, "This line is a comment\n");
    fprintf(fp, "ASCII\n");
    fprintf(fp, "DATASET UNSTRUCTURED_GRID\n");
}

void cigma::VtkIO::write_points(double *points, int nno, int nsd)
{
    fprintf(fp, "POINTS %d float\n", nno);
    for (int i = 0; i < nno; i++)
    {
        fprintf(fp, " %g", points[nsd*i + 0]);
        fprintf(fp, " %g", points[nsd*i + 1]);
        if (nsd == 3)
            fprintf(fp, " %g\n", points[nsd*i + 2]);
        else
            fprintf(fp, " 0.0\n");
    }
}

void cigma::VtkIO::write_point_data(const char *name, double *data, int nno, int ndim)
{
    fprintf(fp, "POINT_DATA %d\n", nno);
    if (ndim == 1)
    {
        fprintf(fp, "SCALARS %s float 1\n", name);
        fprintf(fp, "LOOKUP_TABLE default\n");
        for (int i = 0; i < nno; i++)
            fprintf(fp, "%g\n", data[i]);
    }
    else if ((ndim == 2) || (ndim == 3))
    {
        fprintf(fp, "VECTORS %s float\n", name);
        for (int i = 0; i < nno; i++)
        {
            fprintf(fp, " %g", data[ndim*i + 0]);
            fprintf(fp, " %g", data[ndim*i + 1]);
            if (ndim == 3)
                fprintf(fp, " %g\n", data[ndim*i + 2]);
            else
                fprintf(fp, " 0.0\n");
        }
    }
    else if (ndim == 9)
    {
        fprintf(fp, "TENSORS %s float\n", name);
        for (int i = 0; i < nno; i++)
        {
            for (int j = 0; j < ndim; j++)
            {
                fprintf(fp, " %g", data[ndim*i + j]);
            }
            fprintf(fp, "\n");
        }
    }
}

void cigma::VtkIO::write_cells(int *cells, int nel, int ndofs)
{
    fprintf(fp, "CELLS %d %d\n", nel, nel*(1 + ndofs));
    for (int i = 0; i < nel; i++)
    {
        fprintf(fp, " %d", ndofs);
        for (int j = 0; j < ndofs; j++)
            fprintf(fp, " %d", cells[ndofs*i + j]);
        fprintf(fp, "\n");
    }
}

void cigma::VtkIO::write_cell_types(int nsd, int nel, int ndofs)
{
    int vtkType = 0;

    fprintf(fp, "CELL_TYPES %d\n", nel);
    if (nsd == 3)
    {
        switch (ndofs)
        {
        case  4: vtkType = 10; break; // VTK_TETRA=10
        case  8: vtkType = 12; break; // VTK_HEXAHEDRON=12
        case  6: vtkType = 13; break; // VTK_WEDGE=13
        case  5: vtkType = 14; break; // VTK_PYRAMID=14
        case 10: vtkType = 24; break; // VTK_QUADRATIC_TETRA=24
        case 20: vtkType = 25; break; // VTK_QUADRATIC_HEXAHEDRON=25
        }
    }
    else if (nsd == 2)
    {
        switch (ndofs)
        {
        case 3: vtkType =  5; break; // VTK_TRIANGLE=5
        case 4: vtkType =  9; break; // VTK_QUAD=9
        case 6: vtkType = 22; break; // VTK_QUADRATIC_TRIANGLE=22
        case 8: vtkType = 23; break; // VTK_QUADRATIC_QUAD=23
        }
    }

    assert(vtkType > 0);

    for (int i = 0; i < nel; i++)
        fprintf(fp, "%d\n", vtkType);
}

void cigma::VtkIO::write_cell_data(const char *name, double *data, int nel, int ndim)
{
    fprintf(fp, "CELL_DATA %d\n", nel);
    if (ndim == 1)
    {
        fprintf(fp, "SCALARS %s float 1\n", name);
        fprintf(fp, "LOOKUP_TABLE default\n");
        for (int i = 0; i < nel; i++)
            fprintf(fp, "%g\n", data[i]);
    }
    else if ((ndim == 2) || (ndim == 3))
    {
        fprintf(fp, "VECTORS %s float\n", name);
        for (int i = 0; i < nel; i++)
        {
            fprintf(fp, " %g", data[ndim*i + 0]);
            fprintf(fp, " %g", data[ndim*i + 1]);
            if (ndim == 3)
                fprintf(fp, " %g\n", data[ndim*i + 2]);
            else
                fprintf(fp, " 0.0\n");
        }
    }
    else if (ndim == 9)
    {
        fprintf(fp, "TENSORS %s float\n", name);
        for (int i = 0; i < nel; i++)
        {
            for (int j = 0; j < ndim; j++)
            {
                fprintf(fp, " %g", data[ndim*i + j]);
            }
            fprintf(fp, "\n");
        }
    }
}

File: scratch/utils/vtkio.py
#!/usr/bin/env python

import numpy

VTK_TRIANGLE = 5
VTK_QUAD = 9
VTK_TETRA = 10
VTK_HEXAHEDRON = 12

class VtkWriter(object):
    
    def __init__(self):
        self.fp = None
        self.coords = None
        self.connect = None
        self.residuals = None
        self.cell = None

    def set_mesh(self, coords, connect, cell):
        self.coords = coords
        self.connect = connect
        self.cell = cell

    def set_data(self, residuals):
        self.residuals = residuals
        assert residuals.shape[0] == self.connect.shape[0]

    def open(self, filename):
        print("Creating", fp.name)
        self.fp = open(filename, "w")
        self.write_header()

    def close(self):
        self.fp.close()
        self.fp = None

    def write_header(self):
        fp = self.fp
        fp.write("# vtk DataFile Version 3.0\n")
        fp.write("This line is a comment\n")
        fp.write("ASCII\n")
        fp.write("DATASET UNSTRUCTURED_GRID\n")

    def write_points(self):
        nno,nsd = self.coords.shape
        fp.write("POINTS %d float\n" % nno)
        for n in xrange(nno):
            fp.write("%g %g %g\n" % tuple(coords[n]))

    def write_cells(self):
        fp = self.fp
        connect = self.connect
        nel,ndofs = connect.shape
        
        cellstr = str(ndof) + (" %d" * ndof) + "\n"
        cellstr2 = "%d\n" % cell

        fp.write("CELLS %d %d\n" % (nel, nel + nel*ndof))
        for e in xrange(nel):
            fp.write(cellstr % tuple(connect[e]))
        
        fp.write("CELL_TYPES %d\n" % nel)
        for e in xrange(nel):
            fp.write(cellstr2)
    
    def write_celldata(self, name='epsilon'):
        fp = self.fp
        celldata = self.residuals
        (nel,rank) = celldata.shape
        fp.write("CELL_DATA %d\n" % nel)
        fp.write("SCALARS %s float\n" % name)
        fp.write("LOOKUP_TABLE default\n")
        for e in xrange(nel):
            fp.write("%g\n" % errors[e])


if __name__ == '__main__':

    import numpy
    import tables

    base = 'residuals'
    meshfile = 'cube.h5'

    fp1 = tables.openFile(meshfile, 'r')
    coords = fp1.root.coordinates
    connect = fp1.root.connectivity

    fp2 = tables.openFile(base + '.h5', 'r')
    residuals = fp2.root.epsilon

    writer = VtkWriter()
    writer.open(base + '.vtk')
    writer.set_mesh(coords, connect, VTK_TETRA)
    writer.set_data(residuals)
    writer.write_header()
    writer.write_points()
    writer.write_cells()
    writer.write_celldata()
    writer.close()
